File: backend/src/main/java/com/snackoverflow/toolgether/global/chat/redis/RedisConfig.java
package com.snackoverflow.toolgether.global.chat.redis;

import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.data.redis.connection.RedisConnectionFactory;
import org.springframework.data.redis.connection.RedisStandaloneConfiguration;
import org.springframework.data.redis.connection.lettuce.LettuceConnectionFactory;
import org.springframework.data.redis.core.RedisTemplate;
import org.springframework.data.redis.listener.ChannelTopic;
import org.springframework.data.redis.listener.RedisMessageListenerContainer;
import org.springframework.data.redis.serializer.Jackson2JsonRedisSerializer;
import org.springframework.data.redis.serializer.StringRedisSerializer;

@Slf4j
@Configuration
public class RedisConfig {

    @Value("${spring.data.redis.host}") // application.yml에서 정의된 Redis 호스트
    private String redisHost;

    @Value("${spring.data.redis.port}") // application.yml에서 정의된 Redis 포트
    private int redisPort;

    @Bean
    public RedisConnectionFactory redisConnectionFactory() {
        // RedisStandaloneConfiguration을 사용해 도커 컨테이너의 Redis와 연결 설정
        RedisStandaloneConfiguration redisConfig = new RedisStandaloneConfiguration();
        redisConfig.setHostName(redisHost);
        log.info("redis 연결 시작");
        redisConfig.setPort(redisPort);
        return new LettuceConnectionFactory(redisConfig); // Lettuce를 사용해 Connection Factory 생성
    }

    @Bean
    public RedisTemplate<String, Object> redisTemplate() {
        // RedisTemplate 설정
        RedisTemplate<String, Object> redisTemplate = new RedisTemplate<>();
        redisTemplate.setConnectionFactory(redisConnectionFactory());
        // Key와 Value를 각각 직렬화
        redisTemplate.setKeySerializer(new StringRedisSerializer());
        redisTemplate.setValueSerializer(new Jackson2JsonRedisSerializer<>(Object.class));
        return redisTemplate;
    }

    @Bean
    public ChannelTopic gangnamTopic() {
        return new ChannelTopic("chatroom:gangnam");
    }

    @Bean
    public ChannelTopic mapoTopic() {
        return new ChannelTopic("chatroom:mapo");
    }

    @Bean
    public ChannelTopic seongdongTopic() {
        return new ChannelTopic("chatroom:seongdong");
    }

    @Bean
    public ChannelTopic nowonTopic() {
        return new ChannelTopic("chatroom:nowon");
    }

    @Bean
    public ChannelTopic gwanakTopic() {
        return new ChannelTopic("chatroom:gwanak");
    }

    @Bean
    public RedisMessageListenerContainer redisMessageListenerContainer(RedisConnectionFactory connectionFactory) {
        RedisMessageListenerContainer container = new RedisMessageListenerContainer();
        container.setConnectionFactory(connectionFactory);
        return container;
    }
}

File: backend/src/main/java/com/snackoverflow/toolgether/global/config/RedisConfig.kt
package com.snackoverflow.toolgether.global.config;

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.databind.PropertyNamingStrategies
import com.fasterxml.jackson.module.kotlin.KotlinModule
import com.snackoverflow.toolgether.domain.chat.redis.RedisPubSubEventPublisher
import com.snackoverflow.toolgether.domain.chat.redis.RedisPubSubEventSubscriber
import org.slf4j.Logger
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.data.redis.connection.RedisConnectionFactory;
import org.springframework.data.redis.connection.RedisStandaloneConfiguration;
import org.springframework.data.redis.connection.lettuce.LettuceConnectionFactory;
import org.springframework.data.redis.core.RedisTemplate;
import org.springframework.data.redis.listener.PatternTopic
import org.springframework.data.redis.listener.RedisMessageListenerContainer;
import org.springframework.data.redis.listener.adapter.MessageListenerAdapter
import org.springframework.data.redis.serializer.Jackson2JsonRedisSerializer;
import org.springframework.data.redis.serializer.StringRedisSerializer;

@Configuration
class RedisConfig(
    @Value("\${spring.data.redis.host}") private val redisHost: String,
    @Value("\${spring.data.redis.port}") private val redisPort: Int,
    private val log: Logger
) {

    @Bean
    fun redisConnectionFactory(): RedisConnectionFactory {
        val redisConfig = RedisStandaloneConfiguration().apply {
            hostName = redisHost
            port = redisPort
        }
        log.info("Redis 연결 시작")
        return LettuceConnectionFactory(redisConfig)
    }

    @Bean
    fun redisTemplate(): RedisTemplate<String, Any> {
        return RedisTemplate<String, Any>().apply {
            connectionFactory = redisConnectionFactory()
            keySerializer = StringRedisSerializer()
            valueSerializer = Jackson2JsonRedisSerializer(Any::class.java)
        }
    }

    @Bean
    fun objectMapper(): ObjectMapper = ObjectMapper().apply {

        registerModule(KotlinModule.Builder().build())
        propertyNamingStrategy = PropertyNamingStrategies.SNAKE_CASE
    }

    @Bean
    fun messageListenerAdapter(redisSubscriber: RedisPubSubEventSubscriber): MessageListenerAdapter {
        return MessageListenerAdapter(redisSubscriber, "onMessage")
    }

    @Bean
    fun redisMessageListenerContainer(
        connectionFactory: RedisConnectionFactory,
        listenerAdapter: MessageListenerAdapter
    ): RedisMessageListenerContainer {

        return RedisMessageListenerContainer().apply {
            setConnectionFactory(connectionFactory)
            val topics = listOf(
                PatternTopic("${RedisPubSubEventPublisher.CHAT_EVENT_PREFIX}*"),
                PatternTopic("${RedisPubSubEventPublisher.NOTIFICATION_PREFIX}*"),
                PatternTopic("${RedisPubSubEventPublisher.COMMUNITY_EVENTS_CHANNEL}*"),
            )
            addMessageListener(listenerAdapter, topics)
        }
    }
}